File: tests/Emphasize.UnitTests/EmphasisParserTests.vb
Public Class EmphasisParserTests

    Public Class ParseMethod

        <Fact()>
        Public Sub ReturnsEmptyCollectionForEmptyText()
            Dim target As EmphasisParser
            Dim results As IEnumerable(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse("")

            Assert.Empty(results)
        End Sub


        <Fact()>
        Public Sub ReturnsEmptyCollectionForWhitespace()
            Dim target As EmphasisParser
            Dim results As IEnumerable(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse("    ")

            Assert.Empty(results)
        End Sub


        <Fact()>
        Public Sub ReturnsEmptyCollectionWhenThereAreNoMarkers()
            Dim target As EmphasisParser
            Dim results As IEnumerable(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse("this has no markers")

            Assert.Empty(results)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub ReturnsSingleResultWhenMarkerCoversAllText(
                marker As String,
                type As EmphasisType
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"{marker}all covered{marker}").ToList()

            Assert.Equal(1, results.Count)
            Assert.Equal(0, results(0).StartOffset)
            Assert.Equal(13, results(0).Length)
            Assert.Equal(type, results(0).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub ReturnsSingleResultWhenMarkerIsAtStartOfText(
                marker As String,
                type As EmphasisType
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"{marker}this{marker} is marked").ToList()

            Assert.Equal(1, results.Count)
            Assert.Equal(0, results(0).StartOffset)
            Assert.Equal(6, results(0).Length)
            Assert.Equal(type, results(0).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub ReturnsSingleResultWhenMarkerIsInMiddleOfText(
                marker As String,
                type As EmphasisType
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this {marker}is{marker} marked").ToList()

            Assert.Equal(1, results.Count)
            Assert.Equal(5, results(0).StartOffset)
            Assert.Equal(4, results(0).Length)
            Assert.Equal(type, results(0).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub ReturnsSingleResultWhenMarkerIsAtEndOfText(
                marker As String,
                type As EmphasisType
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this is {marker}marked text{marker}").ToList()

            Assert.Equal(1, results.Count)
            Assert.Equal(8, results(0).StartOffset)
            Assert.Equal(13, results(0).Length)
            Assert.Equal(type, results(0).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub IgnoresMarkersInMiddleOfWords(
                marker As String,
                type As EmphasisType
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"mi{marker}dd{marker}le").ToList()

            Assert.Equal(0, results.Count)
        End Sub


        <MemberData(NameOf(GetMarkerTypesAndPunctuation))>
        <Theory()>
        Public Sub DetectsEndMarkerBeforePunctuation(
                marker As String,
                type As EmphasisType,
                punctuation As Char
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this is {marker}marked{marker}{punctuation} and has punctuation").ToList()

            Assert.Equal(1, results.Count)
            Assert.Equal(8, results(0).StartOffset)
            Assert.Equal(8, results(0).Length)
            Assert.Equal(type, results(0).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypesAndOpeningBrackets))>
        <Theory()>
        Public Sub DetectsStartMarkerAfterBrackets(
                marker As String,
                type As EmphasisType,
                bracket As Char
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this is {bracket}{marker}marked{marker} with brackets").ToList()

            Assert.Equal(1, results.Count)
            Assert.Equal(9, results(0).StartOffset)
            Assert.Equal(8, results(0).Length)
            Assert.Equal(type, results(0).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypesAndClosingBrackets))>
        <Theory()>
        Public Sub DetectsEndMarkerBeforeBrackets(
                marker As String,
                type As EmphasisType,
                bracket As Char
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this is {marker}marked{marker}{bracket} with brackets").ToList()

            Assert.Equal(1, results.Count)
            Assert.Equal(8, results(0).StartOffset)
            Assert.Equal(8, results(0).Length)
            Assert.Equal(type, results(0).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub ReturnsMultipleResultsWhenTextContainsMultipleMarkedSpans(
                marker As String,
                type As EmphasisType
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this {marker}is{marker} marked {marker}and{marker} so is this").ToList()

            Assert.Equal(2, results.Count)

            Assert.Equal(5, results(0).StartOffset)
            Assert.Equal(4, results(0).Length)
            Assert.Equal(type, results(0).Type)

            Assert.Equal(17, results(1).StartOffset)
            Assert.Equal(5, results(1).Length)
            Assert.Equal(type, results(1).Type)
        End Sub


        <Fact()>
        Public Sub ReturnsMultipleResultsWhenTextContainsSpansOfDifferentTypes()
            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse("this *is* marked _and_ so `is` this").ToList()

            Assert.Equal(3, results.Count)

            Assert.Equal(5, results(0).StartOffset)
            Assert.Equal(4, results(0).Length)
            Assert.Equal(EmphasisType.Bold, results(0).Type)

            Assert.Equal(17, results(1).StartOffset)
            Assert.Equal(5, results(1).Length)
            Assert.Equal(EmphasisType.Italic, results(1).Type)

            Assert.Equal(26, results(2).StartOffset)
            Assert.Equal(4, results(2).Length)
            Assert.Equal(EmphasisType.Code, results(2).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub IgnoresMarkersThatStartMidWord(
                marker As String,
                type As String
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"wo{marker}rd{marker}").ToList()

            Assert.Equal(0, results.Count)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub IgnoresMarkersThatStartWithinWhitespace(
                marker As String,
                type As String
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this {marker} is not{marker} marked").ToList()

            Assert.Equal(0, results.Count)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub IgnoresMarkersThatEndWithinWhitespace(
                marker As String,
                type As String
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this {marker}is not {marker} marked").ToList()

            Assert.Equal(0, results.Count)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub IgnoresMarkersThatStartAndEndWithinWhitespace(
                marker As String,
                type As String
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this {marker} is not {marker} marked").ToList()

            Assert.Equal(0, results.Count)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub DoesNotProduceSpanWhenEndMarkerNotFound(
                marker As String,
                type As String
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"{marker}word").ToList()

            Assert.Equal(0, results.Count)
        End Sub


        <Fact()>
        Public Sub ProducesSpansOfCombinedTypesWhenDifferentTypesAreNested()
            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)
            Dim index As Integer


            target = New EmphasisParser
            results = target.Parse($"this has *_`mixed`_* types").ToList()

            Assert.Equal(5, results.Count)

            Assert.Equal(9, results(index).StartOffset)
            Assert.Equal(1, results(index).Length)
            Assert.Equal(EmphasisType.Bold, results(index).Type)
            index += 1

            Assert.Equal(10, results(index).StartOffset)
            Assert.Equal(1, results(index).Length)
            Assert.Equal(EmphasisType.Bold Or EmphasisType.Italic, results(index).Type)
            index += 1

            Assert.Equal(11, results(index).StartOffset)
            Assert.Equal(7, results(index).Length)
            Assert.Equal(EmphasisType.Bold Or EmphasisType.Italic Or EmphasisType.Code, results(index).Type)
            index += 1

            Assert.Equal(18, results(index).StartOffset)
            Assert.Equal(1, results(index).Length)
            Assert.Equal(EmphasisType.Bold Or EmphasisType.Italic, results(index).Type)
            index += 1

            Assert.Equal(19, results(index).StartOffset)
            Assert.Equal(1, results(index).Length)
            Assert.Equal(EmphasisType.Bold, results(index).Type)
        End Sub


        <Fact()>
        Public Sub ProducesSpansOfCombinedTypesWhenDifferentTypesOverlap()
            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)
            Dim index As Integer


            target = New EmphasisParser
            results = target.Parse("this *has _mixed `span* types_ that *overlap` with* all _possible_ combinations").ToList()
            '                            bbbbbbbbbbbbbbbbbb             bbbbbbbbbbbbbbb
            '                                 iiiiiiiiiiiiiiiiiiii                          iiiiiiiiii
            '                                        cccccccccccccccccccccccccccc

            Assert.Equal(8, results.Count)

            Assert.Equal(5, results(index).StartOffset)
            Assert.Equal(5, results(index).Length)
            Assert.Equal(EmphasisType.Bold, results(index).Type)
            index += 1

            Assert.Equal(10, results(index).StartOffset)
            Assert.Equal(7, results(index).Length)
            Assert.Equal(EmphasisType.Bold Or EmphasisType.Italic, results(index).Type)
            index += 1

            Assert.Equal(17, results(index).StartOffset)
            Assert.Equal(6, results(index).Length)
            Assert.Equal(EmphasisType.Bold Or EmphasisType.Italic Or EmphasisType.Code, results(index).Type)
            index += 1

            Assert.Equal(23, results(index).StartOffset)
            Assert.Equal(7, results(index).Length)
            Assert.Equal(EmphasisType.Italic Or EmphasisType.Code, results(index).Type)
            index += 1

            Assert.Equal(30, results(index).StartOffset)
            Assert.Equal(6, results(index).Length)
            Assert.Equal(EmphasisType.Code, results(index).Type)
            index += 1

            Assert.Equal(36, results(index).StartOffset)
            Assert.Equal(9, results(index).Length)
            Assert.Equal(EmphasisType.Bold Or EmphasisType.Code, results(index).Type)
            index += 1

            Assert.Equal(45, results(index).StartOffset)
            Assert.Equal(6, results(index).Length)
            Assert.Equal(EmphasisType.Bold, results(index).Type)
            index += 1

            Assert.Equal(56, results(index).StartOffset)
            Assert.Equal(10, results(index).Length)
            Assert.Equal(EmphasisType.Italic, results(index).Type)
        End Sub


        <Fact()>
        Public Sub IgnoresUnclosedSpansThatOverlapWithClosedSpans()
            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse("this *has _unclosed spans*").ToList()

            Assert.Equal(1, results.Count)

            Assert.Equal(5, results(0).StartOffset)
            Assert.Equal(21, results(0).Length)
            Assert.Equal(EmphasisType.Bold, results(0).Type)
        End Sub


        <MemberData(NameOf(GetMarkerTypes))>
        <Theory()>
        Public Sub ReturnsCompletedSpansWhenThereAreIncompleteSpans(
                marker As String,
                type As EmphasisType
            )

            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse($"this {marker}comment{marker} is {marker}incomplete").ToList()

            Assert.Equal(1, results.Count)

            Assert.Equal(5, results(0).StartOffset)
            Assert.Equal(9, results(0).Length)
            Assert.Equal(type, results(0).Type)
        End Sub


        <Fact()>
        Public Sub SupportsCStyleComments()
            Dim target As EmphasisParser
            Dim results As List(Of EmphasisSpan)


            target = New EmphasisParser
            results = target.Parse("/* this is a *c-style* comment */").ToList()

            Assert.Equal(1, results.Count)

            Assert.Equal(13, results(0).StartOffset)
            Assert.Equal(9, results(0).Length)
            Assert.Equal(EmphasisType.Bold, results(0).Type)
        End Sub


        Public Shared Iterator Function GetMarkerTypes() As IEnumerable(Of Object())
            Yield {"*", EmphasisType.Bold}
            Yield {"_", EmphasisType.Italic}
            Yield {"`", EmphasisType.Code}
        End Function


        Public Shared Iterator Function GetMarkerTypesAndPunctuation() As IEnumerable(Of Object())
            For Each item In GetMarkerTypes()
                For Each ch In {"."c, ","c, "?"c, "!"c, ":"c, ";"c}
                    Yield item.Concat(New Object() {ch}).ToArray()
                Next ch
            Next item
        End Function


        Public Shared Iterator Function GetMarkerTypesAndOpeningBrackets() As IEnumerable(Of Object())
            For Each item In GetMarkerTypes()
                For Each ch In {"{"c, "<"c, "("c, "["c}
                    Yield item.Concat(New Object() {ch}).ToArray()
                Next ch
            Next item
        End Function


        Public Shared Iterator Function GetMarkerTypesAndClosingBrackets() As IEnumerable(Of Object())
            For Each item In GetMarkerTypes()
                For Each ch In {"}"c, ">"c, ")"c, "]"c}
                    Yield item.Concat(New Object() {ch}).ToArray()
                Next ch
            Next item
        End Function

    End Class

End Class

File: tests/Emphasize.UnitTests/EmphasisParserTests.cs
using System.Collections.Generic;
using System.Linq;
using Xunit;


namespace Emphasize;

public class EmphasisParserTests {

    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    public void ReturnsEmptyCollectionForEmptyText(bool useMarkdownStyle) {
        Assert.Empty(Parse("", useMarkdownStyle));
    }


    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    public void ReturnsEmptyCollectionForWhitespace(bool useMarkdownStyle) {
        Assert.Empty(Parse("    ", useMarkdownStyle));
    }


    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    public void ReturnsEmptyCollectionWhenThereAreNoMarkers(bool useMarkdownStyle) {
        Assert.Empty(Parse("this has no markers", useMarkdownStyle));
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypes))]
    public void ReturnsSingleResultWhenMarkerCoversAllText(string marker, EmphasisType type, bool useMarkdownStyle) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"{marker}all covered{marker}", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 0, Length = 11 + (2 * marker.Length), Type = type }],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypes))]
    public void ReturnsSingleResultWhenMarkerIsAtStartOfText(string marker, EmphasisType type, bool useMarkdownStyle) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"{marker}this{marker} is marked", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 0, Length = 4 + (2 * marker.Length), Type = type }],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypes))]
    public void ReturnsSingleResultWhenMarkerIsInMiddleOfText(string marker, EmphasisType type, bool useMarkdownStyle) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this {marker}is{marker} marked", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 5, Length = 2 + (2 * marker.Length), Type = type }],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypes))]
    public void ReturnsSingleResultWhenMarkerIsAtEndOfText(string marker, EmphasisType type, bool useMarkdownStyle) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this is {marker}marked text{marker}", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 8, Length = 11 + (2 * marker.Length), Type = type }],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkers))]
    public void IgnoresMarkersInMiddleOfWords(string marker, bool useMarkdownStyle) {
        Assert.Empty(Parse($"mi{marker}dd{marker}le", useMarkdownStyle));
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypesAndPunctuation))]
    public void DetectsEndMarkerBeforePunctuation(string marker, EmphasisType type, bool useMarkdownStyle, char punctuation) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this is {marker}marked{marker}{punctuation} and has punctuation", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 8, Length = 6 + (2 * marker.Length), Type = type }],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypesAndOpeningBrackets))]
    public void DetectsStartMarkerAfterBrackets(string marker, EmphasisType type, bool useMarkdownStyle, char bracket) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this is {bracket}{marker}marked{marker} with brackets", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 9, Length = 6 + (2 * marker.Length), Type = type }],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypesAndClosingBrackets))]
    public void DetectsEndMarkerBeforeBrackets(string marker, EmphasisType type, bool useMarkdownStyle, char bracket) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this is {marker}marked{marker}{bracket} with brackets", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 8, Length = 6 + (2 * marker.Length), Type = type }],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypes))]
    public void ReturnsMultipleResultsWhenTextContainsMultipleMarkedSpans(string marker, EmphasisType type, bool useMarkdownStyle) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this {marker}is{marker} marked {marker}and{marker} so is this", useMarkdownStyle);

        Assert.Equal(
            [
                new EmphasisSpan {StartOffset = 5, Length = 2 + (2 * marker.Length), Type = type },
                new EmphasisSpan {StartOffset = 15 + (2 * marker.Length), Length = 3 + (2 * marker.Length), Type = type }
            ],
            results
        );
    }


    [Fact]
    public void ReturnsMultipleResultsWhenTextContainsSpansOfDifferentTypesAndUseMarkdownStyleIsFalse() {
        IEnumerable<EmphasisSpan> results;


        results = Parse("this *is* marked _and_ so `is` this", false);

        Assert.Equal(
            [
                new EmphasisSpan{ StartOffset = 5, Length = 4, Type = EmphasisType.Bold },
                new EmphasisSpan{ StartOffset = 17, Length = 5, Type = EmphasisType.Italic },
                new EmphasisSpan{ StartOffset = 26, Length = 4, Type = EmphasisType.Code }
            ],
            results
        );
    }


    [Fact]
    public void ReturnsMultipleResultsWhenTextContainsSpansOfDifferentTypesAndUseMarkdownStyleIsTrue() {
        IEnumerable<EmphasisSpan> results;


        results = Parse("this *is* marked _and_ so `is` this and **so is** this", true);

        Assert.Equal(
            [
                new EmphasisSpan{ StartOffset = 5, Length = 4, Type = EmphasisType.Italic },
                new EmphasisSpan{ StartOffset = 17, Length = 5, Type = EmphasisType.Italic },
                new EmphasisSpan{ StartOffset = 26, Length = 4, Type = EmphasisType.Code },
                new EmphasisSpan{ StartOffset = 40, Length = 9, Type = EmphasisType.Bold}
            ],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkers))]
    public void IgnoresMarkersThatStartMidWord(string marker, bool useMarkdownStyle) {
        Assert.Empty(Parse($"wo{marker}rd{marker}", useMarkdownStyle));
    }


    [Theory]
    [MemberData(nameof(GetMarkers))]
    public void IgnoresMarkersThatStartWithinWhitespace(string marker, bool useMarkdownStyle) {
        Assert.Empty(Parse($"this {marker} is not{marker} marked", useMarkdownStyle));
    }


    [Theory]
    [MemberData(nameof(GetMarkers))]
    public void IgnoresMarkersThatEndWithinWhitespace(string marker, bool useMarkdownStyle) {
        Assert.Empty(Parse($"this {marker}is not {marker} marked", useMarkdownStyle));
    }


    [Theory]
    [MemberData(nameof(GetMarkers))]
    public void IgnoresMarkersThatStartAndEndWithinWhitespace(string marker, bool useMarkdownStyle) {
        Assert.Empty(Parse($"this {marker} is not {marker} marked", useMarkdownStyle));
    }

    [Theory]
    [MemberData(nameof(GetMarkers))]
    public void DoesNotProduceSpanWhenEndMarkerNotFound(string marker, bool useMarkdownStyle) {
        Assert.Empty(Parse($"{marker}word", useMarkdownStyle));
    }


    [Fact]
    public void ProducesSpansOfCombinedTypesWhenDifferentTypesAreNestedAndUseMarkdownStyleIsFalse() {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this has *_`mixed`_* types", false);

        Assert.Equal(
            [
                new EmphasisSpan { StartOffset = 9, Length = 1, Type = EmphasisType.Bold },
                new EmphasisSpan { StartOffset = 10, Length = 1, Type = EmphasisType.Bold | EmphasisType.Italic },
                new EmphasisSpan { StartOffset = 11, Length = 7, Type = EmphasisType.Bold | EmphasisType.Italic | EmphasisType.Code },
                new EmphasisSpan { StartOffset = 18, Length = 1, Type = EmphasisType.Bold | EmphasisType.Italic },
                new EmphasisSpan { StartOffset = 19, Length = 1, Type = EmphasisType.Bold }
            ],
            results
        );
    }


    [Fact]
    public void ProducesSpansOfCombinedTypesWhenDifferentTypesAreNestedAndUseMarkdownStyleIsTrue() {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this has **_`mixed`_** types", true);

        Assert.Equal(
            [
                new EmphasisSpan { StartOffset = 9, Length = 2, Type = EmphasisType.Bold },
                new EmphasisSpan { StartOffset = 11, Length = 1, Type = EmphasisType.Bold | EmphasisType.Italic },
                new EmphasisSpan { StartOffset = 12, Length = 7, Type = EmphasisType.Bold | EmphasisType.Italic | EmphasisType.Code },
                new EmphasisSpan { StartOffset = 19, Length = 1, Type = EmphasisType.Bold | EmphasisType.Italic },
                new EmphasisSpan { StartOffset = 20, Length = 2, Type = EmphasisType.Bold }
            ],
            results
        );
    }


    [Fact]
    public void ProducesSpansOfCombinedTypesWhenDifferentTypesOverlapAndUseMarkdownStyleIsFalse() {
        IEnumerable<EmphasisSpan> results;


        results = Parse("this *has _mixed `span* types_ that *overlap` with* all _possible_ combinations", false);
        //                    bbbbbbbbbbbbbbbbbb             bbbbbbbbbbbbbbb
        //                         iiiiiiiiiiiiiiiiiiii                          iiiiiiiiii
        //                                cccccccccccccccccccccccccccc

        Assert.Equal(
            [
                new EmphasisSpan { StartOffset = 5, Length = 5, Type = EmphasisType.Bold },
                new EmphasisSpan { StartOffset = 10, Length = 7, Type = EmphasisType.Bold | EmphasisType.Italic },
                new EmphasisSpan { StartOffset = 17, Length = 6, Type = EmphasisType.Bold | EmphasisType.Italic | EmphasisType.Code },
                new EmphasisSpan { StartOffset = 23, Length = 7, Type = EmphasisType.Italic | EmphasisType.Code },
                new EmphasisSpan { StartOffset = 30, Length = 6, Type = EmphasisType.Code },
                new EmphasisSpan { StartOffset = 36, Length = 9, Type = EmphasisType.Bold | EmphasisType.Code },
                new EmphasisSpan { StartOffset = 45, Length = 6, Type = EmphasisType.Bold },
                new EmphasisSpan { StartOffset = 56, Length = 10, Type = EmphasisType.Italic }
            ],
            results
        );
    }


    [Fact]
    public void ProducesSpansOfCombinedTypesWhenDifferentTypesOverlapAndUseMarkdownStyleIsTrue() {
        IEnumerable<EmphasisSpan> results;


        results = Parse("this **has _mixed `span** types_ that **overlap` with** all _possible_ combinations", true);
        //                    bbbbbbbbbbbbbbbbbbbb             bbbbbbbbbbbbbbbbb
        //                          iiiiiiiiiiiiiiiiiiiii                            iiiiiiiiii
        //                                 cccccccccccccccccccccccccccccc

        Assert.Equal(
            [
                new EmphasisSpan { StartOffset = 5, Length = 6, Type = EmphasisType.Bold },
                new EmphasisSpan { StartOffset = 11, Length = 7, Type = EmphasisType.Bold | EmphasisType.Italic },
                new EmphasisSpan { StartOffset = 18, Length = 7, Type = EmphasisType.Bold | EmphasisType.Italic | EmphasisType.Code },
                new EmphasisSpan { StartOffset = 25, Length = 7, Type = EmphasisType.Italic | EmphasisType.Code },
                new EmphasisSpan { StartOffset = 32, Length = 6, Type = EmphasisType.Code },
                new EmphasisSpan { StartOffset = 38, Length = 10, Type = EmphasisType.Bold | EmphasisType.Code },
                new EmphasisSpan { StartOffset = 48, Length = 7, Type = EmphasisType.Bold },
                new EmphasisSpan { StartOffset = 60, Length = 10, Type = EmphasisType.Italic }
            ],
            results
        );
    }


    [Theory]
    [InlineData("*", false)]
    [InlineData("**", true)]
    public void IgnoresUnclosedSpansThatOverlapWithClosedSpans(string marker, bool useMarkdownStyle) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this {marker}has _unclosed spans{marker}", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 5, Length = 19 + (2 * marker.Length), Type = EmphasisType.Bold }],
            results
        );
    }


    [Theory]
    [MemberData(nameof(GetMarkerTypes))]
    public void ReturnsCompletedSpansWhenThereAreIncompleteSpans(string marker, EmphasisType type, bool useMarkdownStyle) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"this {marker}comment{marker} is {marker}incomplete", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 5, Length = 7 + (2 * marker.Length), Type = type }],
            results
        );
    }


    [Theory]
    [InlineData("*", false)]
    [InlineData("**", true)]
    public void SupportsCStyleComments(string marker, bool useMarkdownStyle) {
        IEnumerable<EmphasisSpan> results;


        results = Parse($"/* this is a {marker}c-style{marker} comment */", useMarkdownStyle);

        Assert.Equal(
            [new EmphasisSpan { StartOffset = 13, Length = 7 + (2 * marker.Length), Type = EmphasisType.Bold }],
            results
        );
    }


    public static IEnumerable<TheoryDataRow<string, EmphasisType, bool>> GetMarkerTypes() {
        yield return new("*", EmphasisType.Bold, false);
        yield return new("_", EmphasisType.Italic, false);
        yield return new("`", EmphasisType.Code, false);

        yield return new("*", EmphasisType.Italic, true);
        yield return new("**", EmphasisType.Bold, true);
        yield return new("_", EmphasisType.Italic, true);
        yield return new("`", EmphasisType.Code, true);
    }


    public static TheoryData<string, bool> GetMarkers() {
        return [.. GetMarkerTypes().Select((x) => (x.Data.Item1, x.Data.Item3))];
    }


    public static IEnumerable<TheoryDataRow<string, EmphasisType, bool, char>> GetMarkerTypesAndPunctuation() {
        foreach (var item in GetMarkerTypes()) {
            foreach (char ch in new[] { '.', ',', '?', '!', ':', ';' }) {
                yield return new(item.Data.Item1, item.Data.Item2, item.Data.Item3, ch);
            }
        }
    }


    public static IEnumerable<TheoryDataRow<string, EmphasisType, bool, char>> GetMarkerTypesAndOpeningBrackets() {
        foreach (var item in GetMarkerTypes()) {
            foreach (char ch in new[] { '{', '<', '(', '[' }) {
                yield return new(item.Data.Item1, item.Data.Item2, item.Data.Item3, ch);
            }
        }
    }


    public static IEnumerable<TheoryDataRow<string, EmphasisType, bool, char>> GetMarkerTypesAndClosingBrackets() {
        foreach (var item in GetMarkerTypes()) {
            foreach (char ch in new[] { '}', '>', ')', ']' }) {
                yield return new(item.Data.Item1, item.Data.Item2, item.Data.Item3, ch);
            }
        }
    }


    private static IEnumerable<EmphasisSpan> Parse(string text, bool useMarkdownStyle) {
        return new EmphasisParser().Parse(text, useMarkdownStyle);
    }

}
